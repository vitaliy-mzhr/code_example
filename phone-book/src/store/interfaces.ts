import { AppHomeState } from '../app/Home/interfaces';
import { AppNavState } from '../app/Nav/interfaces';

export interface AppState {
  contactsState: AppHomeState;
  searchState: AppNavState;
}
