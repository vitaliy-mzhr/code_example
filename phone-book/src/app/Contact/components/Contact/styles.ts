import styled from 'styled-components';

const ContactStyles = styled.div`
  margin: .5rem 0;
  
  .card-action {
    width: 100%;
  }
  .phones-container {
    display: flex;
    justify-content: flex-start;
    flex-direction: column;
    align-items: flex-start;
    flex-wrap: nowrap;
    padding: .5rem .5rem .5rem 1rem;
    
    .phone {
      display: flex;
      justify-content: flex-start;
      align-items: center;
      text-decoration: none;
      color: ${props => props.theme.colors.mainColor}
    }
  }
  .phones-header {
    padding: .5rem 1.5rem;
    font-weight: 600;
  }
  .buttons-container {
    position: relative;
    display: flex;
    justify-content: flex-end;
    align-items: center;
  }
  .contact-content {
    display: none;
    padding: 0;
    &.expanded {
      display: block;
    }
    &:last-child {
      padding-bottom: 0;
    }
  }
`;

export default ContactStyles;
