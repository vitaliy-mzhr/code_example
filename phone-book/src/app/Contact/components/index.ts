import Contact from './Contact';
import ContactButtons from './ContactButtons';
import EditContact from './EditContact';

export {
  Contact,
  ContactButtons,
  EditContact,
};
