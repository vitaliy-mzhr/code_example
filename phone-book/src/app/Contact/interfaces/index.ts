import { AppHomeAction } from '../../Home/interfaces';

export interface AppContact {
  id: number;
  name: {
    first: string;
    last: string;
  };
  phone: string[];
  expanded: boolean;
  divider?: string;
}

export interface AppContactDivider {
  divider: string;
}

export interface AppContactAction {
  type: string;
  payload?: any;
}

export interface AppEditContactDispatch {
  fetchContacts: () => AppHomeAction;
  updateContact: (contact: object, id: number) => AppContactAction;
}