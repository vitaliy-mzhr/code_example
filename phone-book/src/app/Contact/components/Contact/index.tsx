import * as React from 'react';
import { AppContact } from '../../interfaces';
import { AppSearchDispatch } from '../../../Nav/interfaces';
import { Card, CardHeader, CardContent, Typography, Divider, CardActionArea } from '@material-ui/core';
import { Phone } from '@material-ui/icons';
import ContactStyles from './styles';
import { ContactButtons } from '../../components';

class Contact extends React.PureComponent<AppContact & AppSearchDispatch, {}> {

  private getHeader(): React.ReactNode {
    const {name, id, toggleContact} = this.props;
    return (
      <CardHeader
        className={`contact-header`}
        title={`${name.first}  ${name.last}`}
        onClick={() => toggleContact(id)}
      />
    );
  }

  public render() {
    const {phone, id, deleteContact, expanded} = this.props;
    return (
      <ContactStyles>
        <Card className="contact-card">
          <CardActionArea className={`card-action`}>
            {this.getHeader()}
          </CardActionArea>
          <Divider/>
          {expanded && <CardContent className={`contact-content expanded`}>
            <Typography component="div">
              <Typography component="div" className={'phones-header'}>
                {phone.length > 1 ? 'Phones:' : 'Phone:'}
              </Typography>
              <div className={`phones-container`}>
                {phone.map((phone: any) => <a href={`tel:${phone}`} className="phone" key={phone}><Phone/>{phone}</a>)}
              </div>
            </Typography>
            <Divider/>
            <div className="buttons-container">
              <ContactButtons contactId={id} deleteContact={deleteContact}/>
            </div>
          </CardContent>}
        </Card>
      </ContactStyles>
    );
  }
}

export default Contact;
