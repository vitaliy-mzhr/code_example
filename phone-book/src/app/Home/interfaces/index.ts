import { AppContact, AppContactAction } from '../../Contact/interfaces';

export interface AppHomeAction {
  type: string;
  payload?: any;
}

export interface AppHomeState {
  isLoading: boolean;
  error: null | string;
  contacts: AppContact[];
}

export interface AppHomeProps {
  isLoading: boolean;
  contacts: AppContact[];
  error?: null | string;
}

export interface AppHomeDispatch {
  fetchContacts: () => AppHomeAction;
  toggleContact: (id: number) => AppContactAction;
  deleteContact: (id: number) => AppContactAction;
}
