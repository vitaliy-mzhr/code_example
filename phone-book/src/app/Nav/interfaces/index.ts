import { AppContact, AppContactAction } from '../../Contact/interfaces';
import { AppHomeAction } from '../../Home/interfaces';

export interface AppNavState {
  inputValue: string;
}

export interface AppSearchProps {
  isLoading: boolean;
  error?: null | string;
  contacts: AppContact[];
  searchInput: string;
}

export interface AppNavProps {
  inputValue: string;
}

export interface AppNavAction {
  type: string;
  payload?: string;
}

export interface AppNavDispatch {
  onSearchChange: (input: string) => AppNavAction;
  clearSearchInput: () => AppNavAction;
  fetchContacts: () => AppHomeAction;
}

export interface AppSearchDispatch {
  toggleContact: (id: number) => AppContactAction;
  deleteContact: (id: number) => AppContactAction;
}