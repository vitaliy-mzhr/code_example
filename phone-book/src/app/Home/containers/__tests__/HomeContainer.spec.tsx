import * as React from 'react';
import ConnectedHomeContainer, { HomeContainer } from '../HomeContainer';
import { shallow } from 'enzyme';
import { AppContact } from '../../../Contact/interfaces';
import { AppHomeDispatch, AppHomeProps } from '../../interfaces';
// @ts-ignore
import configureStore from 'redux-mock-store';

describe('Home container test', () => {
  const mockContacts: AppContact[] = [
    {
      id: 1,
      name: {
        first: 'man',
        last: 'awesome',
      },
      phone: [],
      expanded: false,
    },
    {
      id: 2,
      name: {
        first: 'man',
        last: 'awesome',
      },
      phone: [],
      expanded: false,
    }
  ];

  it('ConnectedHomeContainer should renders without crashing with initial state', () => {
    const initialStore: any =  {
      contactsState: {
        isLoading: false,
        contacts: [],
        error: null,
      },
      fetchContacts: jest.fn(),
      toggleContact: jest.fn(),
      deleteContact: jest.fn(),
    };
    const mockStore = configureStore();
    // @ts-ignore
    expect(shallow(<ConnectedHomeContainer store={mockStore(initialStore)}/>)).toMatchSnapshot();
  });

  it('HomeContainer should renders without crashing with initial state', () => {
    const mockProps: AppHomeProps & AppHomeDispatch = {
      isLoading: false,
      contacts: [],
      error: null,
      fetchContacts: jest.fn(),
      toggleContact: jest.fn(),
      deleteContact: jest.fn(),
    };
    const wrapper = shallow(<HomeContainer {...mockProps}/>);
    const instance = wrapper.instance();
    expect(wrapper).toMatchSnapshot();
    // @ts-ignore
    expect(instance.props.fetchContacts).toBeCalled();
  });

  it('HomeContainer should renders without crashing with contacts', () => {
    const mockProps: AppHomeProps & AppHomeDispatch = {
      isLoading: false,
      contacts: mockContacts,
      error: null,
      fetchContacts: jest.fn(),
      toggleContact: jest.fn(),
      deleteContact: jest.fn(),
    };
    const wrapper = shallow(<HomeContainer {...mockProps}/>);
    const instance: any = wrapper.instance();
    expect(wrapper).toMatchSnapshot();
    expect(instance.props.fetchContacts).not.toBeCalled();
  });

  it('HomeContainer should renders with error', () => {
    const mockProps: AppHomeProps & AppHomeDispatch = {
      isLoading: false,
      contacts: [],
      error: 'error',
      fetchContacts: jest.fn(),
      toggleContact: jest.fn(),
      deleteContact: jest.fn(),
    };
    const wrapper = shallow(<HomeContainer {...mockProps}/>);
    expect(wrapper).toMatchSnapshot();
  });
});