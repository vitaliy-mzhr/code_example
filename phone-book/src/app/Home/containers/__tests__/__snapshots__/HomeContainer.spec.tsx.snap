// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home container test ConnectedHomeContainer should renders without crashing with initial state 1`] = `
<HomeContainer
  contacts={Array []}
  deleteContact={[Function]}
  error={null}
  fetchContacts={[Function]}
  isLoading={false}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  toggleContact={[Function]}
/>
`;

exports[`Home container test HomeContainer should renders with error 1`] = `
<ErrorPage
  error="error"
/>
`;

exports[`Home container test HomeContainer should renders without crashing with contacts 1`] = `
<Home
  contacts={
    Array [
      Object {
        "expanded": false,
        "id": 1,
        "name": Object {
          "first": "man",
          "last": "awesome",
        },
        "phone": Array [],
      },
      Object {
        "expanded": false,
        "id": 2,
        "name": Object {
          "first": "man",
          "last": "awesome",
        },
        "phone": Array [],
      },
    ]
  }
  deleteContact={[MockFunction]}
  isLoading={false}
  toggleContact={[MockFunction]}
/>
`;

exports[`Home container test HomeContainer should renders without crashing with initial state 1`] = `
<Home
  contacts={Array []}
  deleteContact={[MockFunction]}
  isLoading={false}
  toggleContact={[MockFunction]}
/>
`;
